File: src/lib/llm-provider.ts
export type LegacyClientLLMConfig = {
  openrouterApiKey?: string | null;
  openrouterModel?: string | null;
};

export type LLMProvider = "local" | "openrouter";

export type LLMConfig = {
  provider: LLMProvider;
  baseUrl: string;
  endpoint: string;
  model: string;
  apiKey: string;
};

const OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1";
const LOCAL_DEFAULT_BASE_URL = "http://127.0.0.1:11434/v1";
const LOCAL_DEFAULT_MODEL = "qwen2.5:32b";

type ProviderMode = LLMProvider | "schedule";

function normalizeBaseUrl(value: string): string {
  return value.trim().replace(/\/+$/, "");
}

function configuredProvider(): ProviderMode {
  const value = process.env.LLM_PROVIDER?.trim().toLowerCase();
  if (value === "local" || value === "openrouter") return value;
  return "schedule";
}

function localScheduleActive(now: Date): boolean {
  const timezone = process.env.LLM_SCHEDULE_TIMEZONE?.trim() || "America/Argentina/Buenos_Aires";
  const parts = new Intl.DateTimeFormat("en-GB", {
    timeZone: timezone,
    hour: "2-digit",
    minute: "2-digit",
    hourCycle: "h23",
  }).formatToParts(now);
  const hour = Number(parts.find((part) => part.type === "hour")?.value ?? "0");
  const minute = Number(parts.find((part) => part.type === "minute")?.value ?? "0");
  const current = hour * 60 + minute;
  const start = parseScheduleTime(process.env.LLM_LOCAL_START, 9 * 60 + 30);
  const end = parseScheduleTime(process.env.LLM_LOCAL_END, 17 * 60 + 30);
  return start <= end ? current >= start && current < end : current >= start || current < end;
}

function parseScheduleTime(value: string | undefined, fallback: number): number {
  const match = value?.trim().match(/^(?:[01]?\d|2[0-3]):[0-5]\d$/);
  if (!match) return fallback;
  const [hours, minutes] = match[0].split(":").map(Number);
  return hours * 60 + minutes;
}

export function resolveLLMProvider(now = new Date()): LLMProvider {
  const configured = configuredProvider();
  return configured === "schedule" ? (localScheduleActive(now) ? "local" : "openrouter") : configured;
}

export function resolveLLMConfig(client?: LegacyClientLLMConfig | null, now = new Date()): LLMConfig {
  const provider = resolveLLMProvider(now);

  if (provider === "local") {
    const baseUrl = normalizeBaseUrl(process.env.LLM_LOCAL_BASE_URL || process.env.LLM_BASE_URL || LOCAL_DEFAULT_BASE_URL);
    return {
      provider,
      baseUrl,
      endpoint: `${baseUrl}/chat/completions`,
      model: process.env.LLM_LOCAL_MODEL?.trim() || process.env.LLM_MODEL?.trim() || LOCAL_DEFAULT_MODEL,
      apiKey: process.env.LLM_LOCAL_API_KEY?.trim() || process.env.LLM_API_KEY?.trim() || "ollama",
    };
  }

  return resolveOpenRouterConfig(client);
}

export function resolveOpenRouterConfig(client?: LegacyClientLLMConfig | null): LLMConfig {
  const baseUrl = normalizeBaseUrl(process.env.OPENROUTER_BASE_URL || OPENROUTER_BASE_URL);
  return {
    provider: "openrouter",
    baseUrl,
    endpoint: `${baseUrl}/chat/completions`,
    model: client?.openrouterModel?.trim() || process.env.OPENROUTER_MODEL?.trim() || "google/gemini-2.0-flash-lite",
    apiKey: client?.openrouterApiKey?.trim() || process.env.OPENROUTER_API_KEY?.trim() || "",
  };
}

export function llmHeaders(config: LLMConfig, title: string): Record<string, string> {
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (config.apiKey) headers.Authorization = `Bearer ${config.apiKey}`;
  if (config.provider === "openrouter") {
    headers["HTTP-Referer"] = "https://los5apostoles.local";
    headers["X-Title"] = title;
  }
  return headers;
}

export async function fetchChatCompletion(
  config: LLMConfig,
  payload: Record<string, unknown>,
  title: string,
  client?: LegacyClientLLMConfig | null,
  init?: Pick<RequestInit, "signal">,
): Promise<{ response: Response; config: LLMConfig; usedFallback: boolean }> {
  const request = async (activeConfig: LLMConfig) => fetch(activeConfig.endpoint, {
    method: "POST",
    headers: llmHeaders(activeConfig, title),
    body: JSON.stringify({ ...payload, model: activeConfig.model }),
    ...init,
  });

  try {
    const response = await request(config);
    if (response.ok || config.provider !== "local") return { response, config, usedFallback: false };

    const fallback = resolveOpenRouterConfig(client);
    if (!fallback.apiKey) return { response, config, usedFallback: false };
    console.warn(`[LLM] Local HTTP ${response.status}; se usa OpenRouter como respaldo para ${title}.`);
    return { response: await request(fallback), config: fallback, usedFallback: true };
  } catch (error) {
    // Una cancelacion expresa del operador no es una falla del proveedor local.
    // Reintentar contra OpenRouter ignora la intencion de cancelar y duplica la
    // solicitud, incluso cuando el request original ya recibio la signal.
    if (init?.signal?.aborted) throw error;
    if (config.provider !== "local") throw error;
    const fallback = resolveOpenRouterConfig(client);
    if (!fallback.apiKey) throw error;
    console.warn(`[LLM] El proveedor local no respondió; se usa OpenRouter como respaldo para ${title}.`);
    return { response: await request(fallback), config: fallback, usedFallback: true };
  }
}

File: agents/llm_provider.py
"""Shared OpenAI-compatible LLM transport for Python agents."""

from __future__ import annotations

import json
import os
import re
import urllib.error
import urllib.request
from datetime import datetime
from zoneinfo import ZoneInfo
from pathlib import Path


OPENROUTER_BASE_URL = "https://openrouter.ai/api/v1"
LOCAL_DEFAULT_BASE_URL = "http://127.0.0.1:11434/v1"
LOCAL_DEFAULT_MODEL = "qwen2.5:32b"


def load_env() -> None:
    env_path = Path(__file__).resolve().parents[1] / ".env"
    if not env_path.exists():
        return
    for raw_line in env_path.read_text(encoding="utf-8").splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))


def provider() -> str:
    load_env()
    configured = os.getenv("LLM_PROVIDER", "schedule").strip().lower()
    if configured in {"local", "openrouter"}:
        return configured
    return "local" if _local_schedule_active() else "openrouter"


def _parse_time(value: str, fallback: int) -> int:
    match = re.fullmatch(r"(?:[01]?\d|2[0-3]):[0-5]\d", value.strip())
    if not match:
        return fallback
    hours, minutes = map(int, value.split(":"))
    return hours * 60 + minutes


def _local_schedule_active(now: datetime | None = None) -> bool:
    timezone = os.getenv("LLM_SCHEDULE_TIMEZONE", "America/Argentina/Buenos_Aires").strip()
    try:
        current_time = now or datetime.now(ZoneInfo(timezone))
    except Exception:
        current_time = now or datetime.now(ZoneInfo("America/Argentina/Buenos_Aires"))
    current = current_time.hour * 60 + current_time.minute
    start = _parse_time(os.getenv("LLM_LOCAL_START", "09:30"), 9 * 60 + 30)
    end = _parse_time(os.getenv("LLM_LOCAL_END", "17:30"), 17 * 60 + 30)
    return start <= current < end if start <= end else current >= start or current < end


def model(requested: str = "") -> str:
    current_provider = provider()
    if current_provider == "local":
        return os.getenv("LLM_LOCAL_MODEL", "").strip() or os.getenv("LLM_MODEL", LOCAL_DEFAULT_MODEL).strip() or LOCAL_DEFAULT_MODEL
    return requested.strip() or os.getenv("OPENROUTER_MODEL", "").strip() or "google/gemini-2.0-flash-lite"


def config(requested_model: str = "") -> dict[str, str]:
    current_provider = provider()
    if current_provider == "local":
        base_url = os.getenv("LLM_LOCAL_BASE_URL", "").strip() or os.getenv("LLM_BASE_URL", LOCAL_DEFAULT_BASE_URL).strip()
        api_key = os.getenv("LLM_LOCAL_API_KEY", "").strip() or os.getenv("LLM_API_KEY", "ollama").strip() or "ollama"
    else:
        base_url = os.getenv("OPENROUTER_BASE_URL", OPENROUTER_BASE_URL).strip()
        api_key = os.getenv("OPENROUTER_API_KEY", "").strip()
    if not api_key:
        raise RuntimeError(f"Falta API key para el proveedor {current_provider}")
    return {
        "provider": current_provider,
        "base_url": base_url,
        "endpoint": f"{base_url}/chat/completions",
        "model": model(requested_model),
        "api_key": api_key,
    }


def chat(system: str, user: str, requested_model: str = "", temperature: float = 0.3, max_tokens: int | None = None, timeout: int = 120) -> str:
    settings = config(requested_model)
    payload: dict = {
        "model": settings["model"],
        "temperature": temperature,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
    }
    if max_tokens:
        payload["max_tokens"] = max_tokens
    headers = {
        "Authorization": f"Bearer {settings['api_key']}",
        "Content-Type": "application/json",
    }
    if settings["provider"] == "openrouter":
        headers["HTTP-Referer"] = "https://los5apostoles.local/"
        headers["X-Title"] = "Los 5 Apostoles"
    request = urllib.request.Request(
        settings["endpoint"],
        data=json.dumps(payload).encode("utf-8"),
        headers=headers,
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            data = json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"Error {settings['provider']} {exc.code}: {detail}") from exc
    return data["choices"][0]["message"]["content"] or ""


def chat_json(system: str, user: str, requested_model: str = "", temperature: float = 0.3, max_tokens: int | None = None, timeout: int = 120) -> dict:
    text = chat(system, user, requested_model, temperature, max_tokens, timeout).strip()
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?", "", text).strip()
        text = re.sub(r"```$", "", text).strip()
    start, end = text.find("{"), text.rfind("}")
    if start == -1 or end <= start:
        raise ValueError("La IA no devolvio un objeto JSON")
    return json.loads(text[start : end + 1])
